e2e/vmss.go

```diff
@@ -82,55 +80,46 @@ func createVMSSWithPayload(ctx context.Context, customData, cseCmd, vmssName str
 		return nil, fmt.Errorf(" prepare model for VMSS %q: %w", vmssName, err)
 	}
 
-	var pollErr error
-	var vmssResp *armcompute.VirtualMachineScaleSetsClientCreateOrUpdateResponse
-
-	for i := 0; i < maxRetries; i++ {
-		createVMSSCtx, cancel := context.WithTimeout(ctx, vmssClientCreateVMSSPollingTimeout)
-		defer cancel()
-
-		vmssResp, pollErr = pollVMSSOperation(createVMSSCtx, vmssName, pollVMSSOperationOpts{
-			pollUntilDone: &runtime.PollUntilDoneOptions{
-				Frequency: vmssClientCreateVMSSPollInterval,
-			},
-		},
-			func() (Poller[armcompute.VirtualMachineScaleSetsClientCreateOrUpdateResponse], error) {
-				return config.Azure.VMSS.BeginCreateOrUpdate(
-					ctx,
-					*opts.clusterConfig.cluster.Properties.NodeResourceGroup,
-					vmssName,
-					model,
-					nil,
-				)
-			})
-
-		if pollErr == nil {
-			return &vmssResp.VirtualMachineScaleSet, nil
-		}
-		if errors.Is(err, context.DeadlineExceeded) || errors.Is(err, context.Canceled) {
-			return nil, fmt.Errorf("create VMSS %q: %w", vmssName, pollErr)
-		}
-		log.Printf("failed to create VMSS. Retry attempts left: %d", maxRetries-(i+1))
+	operation, err := config.Azure.VMSS.BeginCreateOrUpdate(
+		ctx,
+		*opts.clusterConfig.Model.Properties.NodeResourceGroup,
+		vmssName,
+		model,
+		nil,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("begin create VMSS %q: %w", vmssName, err)
 	}
-	return nil, fmt.Errorf("create VMSS %q: %w", vmssName, pollErr)
+	vmssResp, err := operation.PollUntilDone(ctx, &runtime.PollUntilDoneOptions{
+		Frequency: 10 * time.Second,
```

Comment:
the retry logic here definitely needs a refactor, though we should keep it since we used to normally see transient CRP-related errors occur quite frequently when creating the VMSS

Reply:
What are CRP errors?

I tried to think about errors I'd like to retry on. I could only think about network errors. But it applies to any API call.

Reply:
we also have retries enabled on the VMSS client itself.

Chain retries in two places is going to multiple total attempt amount.

Reply:
Once I chained 3 retries in different parts of a system. Innocent 10 attempts become a 10^3 = 1000 attempts spam.

Reply:
so what's annoying about this is that the azure client does not retry service errors automatically, we need logic on top of that layer in order to determine when we should retry, for example when CRP flakes

CRP = compute resource provider, so basically any error related to VM creation that's server-side, e.g. `AllocationFailed`

Reply:
```
	StatusCodes: []int{
			http.StatusRequestTimeout,      // 408
			http.StatusTooManyRequests,     // 429
			http.StatusInternalServerError, // 500
			http.StatusBadGateway,          // 502
			http.StatusServiceUnavailable,  // 503
			http.StatusGatewayTimeout,      // 504
			http.StatusNotFound,            // 404
		},
```

Maybe an http status code is missing in this list?

Reply:
I think `AllocationFailed` is a bad error to retry on. If we don't have capacity now, it's likely we will get in a minute